Make research data more efficient

lkeegan opened the pull request:
- backend
  - refactor data structure
    - now one row per answer session, with a column for each milestone as well as existing columns
    - less duplication of data than previously and reduces processing needed on the frontend
    - reduces data response from ~80MB to ~2MB
    - resolves https://github.com/ssciwr/mondey/issues/295
  - refactor data construction to use pandas
    - most of the remaining processing time is now spent in sql queries
    - we can consider caching this json response in the future, but ok for now and while testing it's nice to always have up to date data
  - return df.to_json as a custom Response
    - much faster than the default fastapi response as it avoids implicit json->dict->conversions and checks
  - resolves https://github.com/ssciwr/mondey/issues/296
  - add gzip compression for responses if larger than 5kb
    - further reduces data response from ~1.9MB to ~ 90kb
    - resolves https://github.com/ssciwr/mondey/issues/103
  - add logging when constructing data
  - add ms to logging timestamps
- frontend
  - improve plotting
    - add explicit keys to plot data instead of deriving them from first datapoint
      - fixes bug where some group by variables were not plotted if they were missing from the first datapoint
    - upgrade unovis
    - enable interpolation with dashed lines for missing data
    - use linear lines between data points when plotting data instead of smoothed ones
  - research page
    - refactor research processing to use new data format
      - reduces amount of group-by processing needed
  - make spinner only appear after 100ms
  - simplify logic to update data
  - research worker
    - add age:0 item to make age data construction logic a bit less confusing
  - upgrade deps, update openapi-ts config
  - worker improvements: it now downloads the data, reduce data tranfer between worker and page

jmsssc commented:
>return df.to_json as a custom Response
>    much faster than the default fastapi response as it avoids implicit json->dict->conversions and checks

Interesting to read that

>add gzip compression for responses if larger than 
>further reduces data response from ~1.9MB to ~ 90kb
wow! I guess the number of "[", "{", "1", "3", "-9" values repeated together is ideal for gzip

jmsssc reviewed (approved):
Looks good to me, works very rapidly with the sample data we have

lkeegan commented on frontend/src/routes/userLand/research/+page.svelte:
Good idea!

lkeegan commented on frontend/tests/ui/research.spec.ts:
```diff
@@ -11,18 +11,17 @@ function make_research_data(n_answer_sessions = 20, n_milestones = 5) {
+			child_age: sample(1, 72),
+			answer_session_id: answer_session_id,
+			number_of_siblings: sample(0, 5),
+			"early birth": ["yes", "no"][sample(0, 1)],
```
Yes

lkeegan commented:
@jmsssc do you want to take another look at this? (in particular I changed the webworker/ui logic&communication a bit and added a delay to the spinner to avoid it flashing in and out for a couple of ms when the data is updated quickyl)

jmsssc commented:
> @jmsssc do you want to take another look at this? (in particular I changed the webworker/ui logic&communication a bit and added a delay to the spinner to avoid it flashing in and out for a couple of ms when the data is updated quickyl)

Sure I will have a look now

lkeegan commented on mondey_backend/src/mondey_backend/routers/research.py:
```diff
@@ -24,8 +25,9 @@ async def get_research_data(
+        df = await extract_research_data(
             session, user_session, research_group_id_filter
         )
+        return Response(df.to_json(orient="records"), media_type="application/json")
```
good point - for the milestones the column name is `milestone_id_<id>` where id is the unique id, so they should be ok, but for child/user questions it could happen.

a quick fix would just be to add a prefix in the backend data, although that could also be broken by a determined/unfortunate user picking weird names.

I guess the clean solution here would be to use the ids for the column headers like we already do for the milestones, e.g. `user_question_1`, `child_question_3` etc, and additionally provide a mapping from these guaranteed unique column headers to a string to be displayed by the UI.

lkeegan replied:
hmm looking into this our `Question` model doesn't have a `name` field -  I guess we should add one as we did for the milestones, otherwise just using the de text as we do currently might lead to very long column names, apart from the uniqueness issues

jmsssc replied:
True, and yes we could set the same label/variable label like "FK03" the researchers use for questions that way. It would be possible to write a one off script to do that with Python for the existing questions, since they just use

`prefixVariableLabel: VariableName` as the question text. We can extract it out and set the name. I could do this as an issue if you want or do you think it belongs better done here with the other changes here?

lkeegan replied:
Yes that would be great - if you could write a script to add `name` to the existing database (& add this to your data mangling branch), I'll update this PR to add `name` to the question model and to use it

lkeegan replied:
I added `name` and we now use the ids for all columns so no more possibility of accidentally having the same column name twice

jmsssc commented on mondey_backend/src/mondey_backend/routers/research.py:
```diff
@@ -24,8 +25,9 @@ async def get_research_data(
         else:
             research_group_id_filter = current_active_researcher.research_group_id
-        return await extract_research_data(
+        df = await extract_research_data(
```
I happened to be testing on the e2e database with this anyway as I had that set up, and I believe if nobody has made any milestone answering sessions at all, then the code will crash in statistics.py called from this file. Maybe we can check that there are answering sessions before  calling this function?

lkeegan replied:
this should be fixed now

jmsssc reviewed (commented):
Looks good to me; the changes I have noted/suggested would only occur in edge cases, but the potential for duplicate columns to cause issues with the returned (later compressed) might come up in a few cases so I think we should find a way to avoid that causing an error. Whether that is best done through preventing people creating duplicates or prefixing columns to their groups (user question, child question or milestone) plus deduplicating them, or some other approach I am not sure.
